File: 704. Binary Search/binary-search.js
// 704. Binary Search (9/29/2021)
// Runtime: 60 ms (91.93%) Memory: 42.49 MB (93.00%) 

/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number}
 */
function search(nums, target) {
    let left = 0,
        right = nums.length - 1;
    while (left <= right) {
        let mid = Math.floor((left + right) / 2);
        if (nums[mid] === target) return mid;
        if (left === right) break;
        if (nums[mid] < target) left = mid + 1;
        else right = mid - 1;
    }
    return -1;
}

File: 704. Binary Search/binary-search.c
// 704. Binary Search (7/22/54201)
// Runtime: 32 ms (91.95%) Memory: 7.22 MB (21.35%) 

int search (int nums[], int nums_size, int target) {
    int upper = nums_size - 1, lower = 0;
    
    while (lower <= upper) {
        int mid = (upper + lower) / 2;
        
        if (nums[mid] == target) {
            return mid;
        }
        
        if (nums[mid] > target) {
            upper = mid - 1;
        }
        else {
            lower = mid + 1;
        }
    }
    
    return -1;
}
